source/common/network/socket_impl.cc

```diff
@@ -0,0 +1,139 @@
+#include "common/network/socket_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/network/address_impl.h"
+#include "common/network/io_socket_handle_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+IoHandlePtr SocketInterface::socket(Address::SocketType socket_type, Address::Type addr_type,
+                                    Address::IpVersion version) {
+#if defined(__APPLE__) || defined(WIN32)
+  int flags = 0;
+#else
+  int flags = SOCK_NONBLOCK;
+#endif
+
+  if (socket_type == Address::SocketType::Stream) {
+    flags |= SOCK_STREAM;
+  } else {
+    flags |= SOCK_DGRAM;
+  }
+
+  int domain;
+  if (addr_type == Address::Type::Ip) {
+    if (version == Address::IpVersion::v6) {
+      domain = AF_INET6;
+    } else {
+      ASSERT(version == Address::IpVersion::v4);
+      domain = AF_INET;
+    }
+  } else {
+    ASSERT(addr_type == Address::Type::Pipe);
+    domain = AF_UNIX;
+  }
+
+  const Api::SysCallSocketResult result = Api::OsSysCallsSingleton::get().socket(domain, flags, 0);
+  RELEASE_ASSERT(SOCKET_VALID(result.rc_),
+                 fmt::format("socket(2) failed, got error: {}", strerror(result.errno_)));
+  IoHandlePtr io_handle = std::make_unique<IoSocketHandleImpl>(result.rc_);
+
+#if defined(__APPLE__) || defined(WIN32)
+  // Cannot set SOCK_NONBLOCK as a ::socket flag.
+  const int rc = Api::OsSysCallsSingleton::get().setsocketblocking(io_handle->fd(), false).rc_;
+  RELEASE_ASSERT(!SOCKET_FAILURE(rc), "");
+#endif
+
+  return io_handle;
+}
+
+IoHandlePtr SocketInterface::socket(Address::SocketType socket_type,
+                                    const Address::InstanceConstSharedPtr addr) {
+  Address::IpVersion ip_version = addr->ip() ? addr->ip()->version() : Address::IpVersion::v4;
+  IoHandlePtr io_handle = SocketInterface::socket(socket_type, addr->type(), ip_version);
+  if (addr->type() == Address::Type::Ip && addr->ip()->version() == Address::IpVersion::v6) {
+    // Setting IPV6_V6ONLY restricts the IPv6 socket to IPv6 connections only.
+    const int v6only = addr->ip()->v6only();
+    const Api::SysCallIntResult result = Api::OsSysCallsSingleton::get().setsockopt(
+        io_handle->fd(), IPPROTO_IPV6, IPV6_V6ONLY, reinterpret_cast<const char*>(&v6only),
+        sizeof(v6only));
+    RELEASE_ASSERT(!SOCKET_FAILURE(result.rc_), "");
+  }
+  return io_handle;
+}
+
+bool SocketInterface::ipFamilySupported(int domain) {
```

Comment:
This is a dup of the function in the address file, right? Can we collapse this?

Reply:
Yup, it tries to reproduce what Ipv6Instance::socket() was doing. I think that the v6 specific logic should eventually be handled with socket->setOption(setv6only) so this could be cleaned up. However, v6only is true by default, so that could mean a lot of changes.

Reply:
Sorry what I mean is this is the same function as https://github.com/envoyproxy/envoy/pull/11278/files#diff-3a2eb9279fdb086a43b5b47dcbe278cdR22-R32, right? Can you just have this be a utility function in network::utility or something?

Reply:
Sorry, I completely missed that you were pointing at ipFamilySupported. 

I was about to mark this as done, but then noticed that the socket functions were built such that they cannot fail (RELEASE_ASSERT on the socket() return value), therefore we can't rely on this logic in the utility function to test if a socket type is usable. 

So we're left with three options:
- change SocketInterface::socket() to allow for failures (not sure about the impact in the rest of the codebase ..)
- do the af test in a different way, but that will probably involve introducing a new SocketInterface function, i.e., something like SocketInterface::socket() that can fail.
- keep on using this for now ...

Reply:
Sorry I'm really confused. AFAICT the body of this function is exactly the same as the one I pointed to. Why can't they be shared?

Reply:
As usual, I'm the one confused (I thought you felt this function didn't belong to SocketInterface).

The duplication is actually a byproduct of how I split the original PR. In the next PR, I'll completely remove the validateIpvXSupported check from the address, together with the remaining socket api calls, and wanted to have that conversation once we get there. If you think we shouldn't remove the validations, I can add a utility function now.

Reply:
OK if you are going to deal with this in a follow up it's fine to leave. SGTM. Can you push the other change and I can take a look?

Reply:
Here's a draft [patch](https://github.com/florincoras/envoy/commit/eadbfbca266605fb95f1d254ad2790a1addf6900) 

I can also rebase the patch and push a separate PR if that works better.

Reply:
We can just do it as a follow up. Thank you!

Reply:
Thank you for the quick replies! And apologies again for the confusion.